File: clients/typescript/test/scenarios.test.ts
// The shared behaviour suite (clients/contract/scenarios.json, ADR 0009).
import { test, before, after } from "node:test";
import assert from "node:assert/strict";
import { spawn, type ChildProcess } from "node:child_process";
import { readFileSync } from "node:fs";
import { createInterface } from "node:readline";
import { inspect } from "node:util";
import * as nm from "../src/index.js";
import { DISPATCH } from "./dispatch.js";

// Compiled into dist-test/test/, so clients/ is three levels up.
const CONTRACT = new URL("../../../contract/", import.meta.url);
const SCEN = JSON.parse(
  readFileSync(new URL("scenarios.json", CONTRACT), "utf8")
);
const TOKEN: string = SCEN.token;
let proc: ChildProcess;
let url = "";
let closed = "";

before(async () => {
  proc = spawn("sh", ["scenario-server.sh", "-scenarios", "scenarios.json"], {
    cwd: CONTRACT,
    stdio: ["ignore", "pipe", "inherit"],
  });
  const line: string = await new Promise((res) =>
    createInterface({ input: proc.stdout! }).once("line", res)
  );
  const parts = line.split(" ");
  url = parts[1];
  closed = parts[2].split("=")[1];
});
after(() => {
  proc.kill();
});

function isEmpty(r: unknown): boolean {
  return (
    (Array.isArray(r) && r.length === 0) ||
    (typeof r === "object" &&
      r !== null &&
      Array.isArray((r as any).results) &&
      (r as any).results.length === 0)
  );
}

function classify(r: unknown, e: unknown): string {
  if (e === undefined) return isEmpty(r) ? "empty" : "ok";
  if (!(e instanceof nm.NovamemError)) throw e;
  if (e.canceled) return "canceled";
  if (e.unavailable) return "unavailable";
  if (e.notFound) return "not_found";
  return "error";
}

function subset(want: unknown, got: unknown, path = "$"): string {
  if (want !== null && typeof want === "object" && !Array.isArray(want)) {
    if (got === null || typeof got !== "object")
      return `${path}: want an object, got ${JSON.stringify(got)}`;
    for (const [k, v] of Object.entries(want)) {
      const d = subset(v, (got as any)[k], `${path}.${k}`);
      if (d) return d;
    }
    return "";
  }
  return JSON.stringify(want) === JSON.stringify(got)
    ? ""
    : `${path}: got ${JSON.stringify(got)}, want ${JSON.stringify(want)}`;
}

// proved by: removing the degraded-empty check in Client.search fails
// search-degraded-empty-is-unavailable; removing the token redaction fails
// token-echoed-in-401-is-redacted.
for (const s of SCEN.scenarios) {
  test(s.id, async () => {
    const call = s.call;
    let r: unknown;
    let e: unknown;
    if (call.class === "ctor") {
      try {
        new nm.Client({
          baseUrl: call.args.baseUrl.replace("<server>", url),
          token: call.args.token,
        });
      } catch (err) {
        e = err;
      }
    } else {
      const baseUrl = JSON.stringify(s.respond).includes('"refused"')
        ? `http://127.0.0.1:${closed}/s/${s.id}`
        : `${url}/s/${s.id}`;
      const o = { baseUrl, token: TOKEN, timeoutMs: SCEN.timeoutMs };
      const clients = {
        Client: new nm.Client(o),
        Management: new nm.Management(o),
        Admin: new nm.Admin(o),
      };
      const ac = new AbortController();
      const timer = call.cancelAfterMs
        ? setTimeout(() => ac.abort(), call.cancelAfterMs)
        : undefined;
      try {
        r = await DISPATCH[call.method](clients, call.args, ac.signal);
      } catch (err) {
        e = err;
      } finally {
        clearTimeout(timer);
      }
    }
    const exp = s.expect;
    assert.equal(classify(r, e), exp.outcome, `${s.id}: ${String(e)}`);
    if (e instanceof nm.NovamemError) {
      assert.ok(!String(e).includes(TOKEN), "token leaked into String(err)");
      assert.ok(!inspect(e).includes(TOKEN), "token leaked into inspect(err)");
      assert.ok(
        !JSON.stringify(e).includes(TOKEN),
        "token leaked into JSON.stringify(err)"
      );
      if ("retryable" in exp) assert.equal(e.retryable, exp.retryable);
      if ("statusCode" in exp) assert.equal(e.statusCode, exp.statusCode);
      if ("code" in exp) assert.equal(e.code, exp.code);
    }
    if (e instanceof Error && "messageContains" in exp) {
      assert.ok(
        String(e).toLowerCase().includes(exp.messageContains.toLowerCase()),
        `message ${String(e)}`
      );
    }
    if ("result" in exp) assert.equal(subset(exp.result, r), "");
    if (call.class !== "ctor") {
      const v = await (await fetch(`${url}/_verdict/${s.id}`)).json();
      assert.deepEqual(v.mismatches, []);
      if (Array.isArray(s.expectRequest) && s.expectRequest.length === 0)
        assert.equal(v.requests, 0);
    }
  });
}

File: clients/ruby/test/scenarios_test.rb
# frozen_string_literal: true

# The shared behaviour suite (clients/contract/scenarios.json, ADR 0009).

require "minitest/autorun"
require "json"
require "net/http"
require "novamem"
require_relative "dispatch"

CONTRACT = File.expand_path("../../contract", __dir__)
SCEN = JSON.parse(File.read(File.join(CONTRACT, "scenarios.json")))
TOKEN = SCEN.fetch("token")

class ScenariosTest < Minitest::Test
  def self.server
    @server ||= begin
      io = IO.popen(["sh", "scenario-server.sh", "-scenarios", "scenarios.json"], chdir: CONTRACT)
      _, url, closed = io.gets.split
      Minitest.after_run { Process.kill("KILL", io.pid) }
      [url, closed.split("=").last]
    end
  end

  def empty?(r)
    w = r.respond_to?(:to_h_wire) ? r.to_h_wire : r
    w == [] || (w.is_a?(Hash) && w["results"] == [])
  end

  def classify(r, e)
    return (empty?(r) ? "empty" : "ok") if e.nil?
    return "unavailable" if e.is_a?(Novamem::Error) && e.unavailable?
    return "not_found" if e.is_a?(Novamem::Error) && e.not_found?

    "error"
  end

  def subset(want, got, path = "$")
    return (got == want ? nil : "#{path}: got #{got.inspect}, want #{want.inspect}") unless want.is_a?(Hash)
    return "#{path}: want an object, got #{got.inspect}" unless got.is_a?(Hash)

    want.each { |k, v| (d = subset(v, got[k], "#{path}.#{k}")) && (return d) }
    nil
  end

  # proved by: removing the degraded-empty check in Client#search fails
  # search-degraded-empty-is-unavailable; removing the redaction fails
  # token-echoed-in-401-is-redacted.
  SCEN.fetch("scenarios").each do |s|
    define_method("test_#{s["id"].tr("-", "_")}") do
      skip "the Ruby client has no cancellation primitive: #{s["id"]}" if Array(s["requires"]).include?("cancel")
      url, closed = self.class.server
      call = s["call"]
      r = e = nil
      if call["class"] == "ctor"
        begin
          Novamem::Client.new(call["args"]["baseUrl"].sub("<server>", url), call["args"]["token"])
        rescue Novamem::ConfigError, Novamem::Error => err
          e = err
        end
      else
        refused = s["respond"].to_json.include?('"refused"')
        base = refused ? "http://127.0.0.1:#{closed}/s/#{s["id"]}" : "#{url}/s/#{s["id"]}"
        opts = { timeout: SCEN["timeoutMs"] / 1000.0 }
        clients = { Client: Novamem::Client.new(base, TOKEN, **opts),
                    Management: Novamem::Management.new(base, TOKEN, **opts),
                    Admin: Novamem::Admin.new(base, TOKEN, **opts) }
        begin
          r = DISPATCH.fetch(call["method"]).call(clients, call["args"])
        rescue Novamem::Error => err
          e = err
        end
      end
      exp = s["expect"]
      assert_equal exp["outcome"], classify(r, e), "#{s["id"]}: #{e.inspect}"
      if e
        refute_includes e.message, TOKEN
        refute_includes e.inspect, TOKEN
        if e.is_a?(Novamem::Error)
          assert_equal exp["retryable"], e.retryable?, "retryable" if exp.key?("retryable")
          assert_equal exp["statusCode"], e.status_code, "statusCode" if exp.key?("statusCode")
          assert_equal exp["code"], e.code, "code" if exp.key?("code")
        end
        assert_includes e.message.downcase, exp["messageContains"].downcase if exp.key?("messageContains")
      end
      if exp.key?("result")
        got = r.respond_to?(:to_h_wire) ? r.to_h_wire : r
        assert_nil subset(exp["result"], got)
      end
      next if call["class"] == "ctor"

      v = JSON.parse(Net::HTTP.get(URI("#{url}/_verdict/#{s["id"]}")))
      assert_equal [], v["mismatches"]
      assert_equal 0, v["requests"] if s["expectRequest"] == []
    end
  end
end
